### PR TITLE
JAMES-3693 Provide an API for rate limiting

### PR DESCRIPTION
**Pull request by chibenwa**
Implement memory version for it using the RateLimitJ library. Provide
a contract test running against this implementation.

### CODE REVIEW

File: server/mailet/rate-limiter/src/main/scala/org/apache/james/rate/limiter/api/RateLimiter.scala [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+trait RateLimiter {
+    def rateLimit(key: RateLimitingKey, increaseQuantity: Quantity): Publisher[RateLimitingResult]
+}
+
+trait RateLimiterFactory {
+   def withSpecification(rules: Rules): RateLimiter
+}
+
+sealed trait RateLimitingResult
+case object RateExceeded extends RateLimitingResult
```
Comment (vttranlina) [inline server/mailet/rate-limiter/src/main/scala/org/apache/james/rate/limiter/api/RateLimiter.scala:63]: Should we need `RateLimitingKey` info here?

`case class RateExceeded(key: RateLimitingKey) extends RateLimitingResult`

File: server/mailet/rate-limiter/src/test/scala/org/apache/james/rate/limiter/memory/MemoryRateLimiterTest.scala [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.rate.limiter.memory
+
+import java.time.Duration
+
+import org.apache.james.rate.limiter.api.{RateLimiterContract, RateLimiterFactory}
+
+class MemoryRateLimiterTest extends RateLimiterContract{
```
Comment (vttranlina) [inline server/mailet/rate-limiter/src/test/scala/org/apache/james/rate/limiter/memory/MemoryRateLimiterTest.scala:26]: space
```suggestion
class MemoryRateLimiterTest extends RateLimiterContract {
```

File: server/mailet/rate-limiter/src/test/scala/org/apache/james/rate/limiter/api/RateLimiterContract.scala [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  val rateLimiter = testee().withSpecification(rules)
+
+  SMono(rateLimiter.rateLimit(TestKey("key1"), 2)).block()
+  sleep(Duration.ofSeconds(3))
+  val actual: RateLimitingResult = SMono(rateLimiter.rateLimit(TestKey("key1"), 3)).block()
+  assertThat(actual).isEqualTo(AcceptableRate)
+ }
+
+ @Test
+ def rateLimitingShouldBePartitionnedByKey(): Unit = {
```
Comment (vttranlina) [inline server/mailet/rate-limiter/src/test/scala/org/apache/james/rate/limiter/api/RateLimiterContract.scala:99]: rateLimitingShouldBePartitionedByKey

File: server/mailet/rate-limiter/src/main/scala/org/apache/james/rate/limiter/api/RateLimiter.scala (hunk 1 above)
Comment (chibenwa) [reply server/mailet/rate-limiter/src/main/scala/org/apache/james/rate/limiter/api/RateLimiter.scala:63]: I don't think: the caller would already know the key...

File: server/mailet/rate-limiter/src/test/scala/org/apache/james/rate/limiter/api/RateLimiterContract.scala (hunk 3 above)
Comment (chibenwa) [inline server/mailet/rate-limiter/src/test/scala/org/apache/james/rate/limiter/api/RateLimiterContract.scala:99]:
```suggestion
 def rateLimitingShouldBePartitionedByKey(): Unit = {
```

File: server/mailet/rate-limiter/src/main/scala/org/apache/james/rate/limiter/memory/MemoryRateLimiter.scala [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+class MemoryRateLimiterFactory() extends RateLimiterFactory {
+  override def withSpecification(rules: Rules): RateLimiter = {
+    MemoryRateLimiter(new InMemorySlidingWindowRequestRateLimiter(rules.rules.map(convert).toSet.asJava))
+  }
+
+  private def convert(rule: Rule): RequestLimitRule = RequestLimitRule.of(rule.duration, rule.quantity.value)
+}
+
+case class MemoryRateLimiter(limiter: RequestRateLimiter) extends RateLimiter {
+  override def rateLimit(key: RateLimitingKey, increaseQuantity: Quantity): Publisher[RateLimitingResult] = {
```
Comment (Arsnael) [inline server/mailet/rate-limiter/src/main/scala/org/apache/james/rate/limiter/memory/MemoryRateLimiter.scala:40]: No need of the brackets here

**Comment by Arsnael** [comment]
Rebase to solve conflict

**Review by Arsnael — APPROVED** [review]

**Review by vttranlina — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Pull request merged**